### PR TITLE
[FIXED JENKINS-42762] Allow multiple conditions in when directly

### PR DESCRIPTION
**Pull request by abayer**
* JENKINS issue(s):
    * [JENKINS-42762](https://issues.jenkins-ci.org/browse/JENKINS-42762)
* Description:
    * Reverts changes made in JENKINS-41185 fix. Does not affect the core of the changes in JENKINS-41185 - nested `when` conditions still work exactly the same, we're just back to allowing multiple top-level `when` conditions, where if any of them fail/return false, the `when` is not satisfied. This came to light when [a user reported the regression](https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/125#issuecomment-286322236) so I consider it a priority.
* Documentation changes:
    * n/a
* Users/aliases to notify:
    * @reviewbybees 
    * @rsandell 
    * @kzantow due to `ast-schema.json` reversion.

### CODE REVIEW

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/parser/ModelParser.groovy [hunk 1]
```diff
@@ -121,7 +121,7 @@ class ModelParser implements Parser {
         }
 
         if (pst==null) {
-            // Check if there's a 'pipeline' step somewhere children within the other statements and error out if that's the case.
```
Comment (bitwiseman) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/parser/ModelParser.groovy]: nested

Comment (abayer) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/parser/ModelParser.groovy]: dangit.

**Comment by abayer** [comment]
...yes. You did. =)

File: pipeline-model-definition/src/test/resources/json/whenEnvFalse.json [hunk 2]
```diff
@@ -31,25 +31,25 @@
...
-            "value": {
+            "value":             {
               "isLiteral": true,
               "value": "FOO"
             }
           },
           {
             "key": "value",
-            "value": {
+            "value":             {
```
Comment (bitwiseman) [inline pipeline-model-definition/src/test/resources/json/whenEnvFalse.json]: What's with all the extra spaces?

Comment (abayer) [inline pipeline-model-definition/src/test/resources/json/whenEnvFalse.json]: I copy-pasted the JSON from our internal `ModelASTElement.toJSON().toString(2)` method, and sometimes that comes out weird in re: spacing.

**Comment by IdanAdar** [comment]
Could someone please provide a declarative pipeline example with multiple `when`s?

**Comment by rsandell** [comment]
@IdanAdar https://github.com/jenkinsci/pipeline-model-definition-plugin/blob/master/pipeline-model-definition/src/test/resources/whenNestedCombinations.groovy

**Comment by IdanAdar** [comment]
Thanks @rsandell. Would this then be a valid use?

```
stage ('....') {
    // We're in a PR, lets do something
    when {
        expression {
            (BRANCH_NAME != "develop") || (BRANCH_NAME != "master")
        }
    }

    steps {
        ...
    }
}
```

**Comment by rsandell** [comment]
@IdanAdar yes it is valid, but would trigger for `master` since that is != `develop` and `BRANCH_NAME != "develop"` is evaluated first.
Better expressed would be:

```
 when {
    allOf {
        not { branch "develop" }
        not { branch "master" }
    }
}
```

**Comment by abayer** [comment]
Or
```groovy
when {
    not {
        anyOf {
            branch "develop"
            branch "master"
        }
    }
}
```
because hey. =)

**Comment by IdanAdar** [comment]
So likewise on the positive? 

```
when {
    branch "develop"
}

steps {
    ...
}
```

No need for `expression`?

**Comment by IdanAdar** [comment]
@abayer @rsandell,

This succeeds: 

```
stage ("...") {
         environment {
            JENKINSBOT = credentials('...')
         }
         
         when {
            not { 
               anyOf {
                  branch "develop"
                  branch "master"
               }
            }
         }
         
         steps {
            // sh that uses $JENKINSBOT_PSW
         }
}
```

But this fails... I guess because of the `expression` bit:

```
stage ("...") {
         environment {
            JENKINSBOT = credentials('...')
         }
         
         when {
            allOf {
               expression { return env.shouldBuild != "false"  }
               not { branch "develop" }
               not { branch "master" }
            }
         }
         
         steps {
            // sh using $JENKINSBOT_PSW
         }
}
```

`env.shouldBuild` is defined like this:

```
pipeline {
    parameters {
        booleanParam(defaultValue: true, description: 'Execute pipeline?', name: 'shouldBuild')
   }
   ...
```

And updated like this:

```
result = sh (script: "git log -1 | grep '.*\\[ci skip\\].*'", returnStatus: true) 
if ((result == 0) && (BRANCH_NAME == "develop" || BRANCH_NAME == "master")) {
    echo ("This build should be skipped. Aborting.")
    env.shouldBuild = "false"
}
```

Background: I am doing this in order to skip all stages in some situations, because there is no ci skip functionality that works in declarative pipeline/multibranch pipeline.

**Pull request merged**